File: src/011/main.py
from functools import reduce
from sys import argv

SIZE = 4
with open (argv [1]) as file:
	lines = file.readlines ()
	GRID = list (map (lambda x: list (map (int, x.split ())), lines))

WIDTH = len (GRID [0])
HEIGHT = len (GRID)

largest = 0

def product (array):
	global largest
	largest = max (largest, reduce (lambda a, b: a * b, array))

# horizontal
for i in range (HEIGHT):
	for j in range (WIDTH - SIZE + 1):
		product (GRID[i][j + k] for k in range (SIZE))

# vertical
for i in range (HEIGHT - SIZE + 1):
	for j in range (WIDTH):
		product (GRID[i + k][j] for k in range (SIZE))

# diagonals
for i in range (HEIGHT - SIZE + 1):
	for j in range (WIDTH - SIZE + 1):
		product (GRID[i + k][SIZE - 1 + j - k] for k in range (SIZE))
		product (GRID[i + k][j + k] for k in range (SIZE))

print (largest)

File: src/011/main.rb

SIZE = 4
GRID = (File.readlines ARGV.first).map do |line|
	line.split.map &:to_i
end

WIDTH = GRID.first.length
HEIGHT = GRID.length

$largest = 0

def product &block
	$largest = [$largest, ((SIZE.times.map &block).reduce :*)].max
end

# horizontal
HEIGHT.times do |i|
	(WIDTH - SIZE.pred).times do |j|
		product {|k| GRID[i][j + k]}
	end
end

# vertical
(HEIGHT - SIZE.pred).times do |i|
	WIDTH.times do |j|
		product {|k| GRID[i + k][j]}
	end
end

# diagonals
(HEIGHT - SIZE.pred).times do |i|
	(WIDTH - SIZE.pred).times do |j|
		product {|k| GRID[i + k][SIZE.pred + j - k]}
		product {|k| GRID[i + k][j + k]}
	end
end

puts $largest

